.golangci.yml

```diff
@@ -0,0 +1,282 @@
+    # it's a comma-separated list of prefixes
+    local-prefixes: github.com/m3db/m3
+  gocyclo:
+    # minimal code complexity to report, 30 by default (but we recommend 10-20)
+    min-complexity: 10
+  maligned:
+    # print struct with more effective memory layout or not, false by default
+    suggest-new: true
+  dupl:
+    # tokens count to trigger issue, 150 by default
+    threshold: 100
+  goconst:
+    # minimal length of string constant, 3 by default
+    min-len: 3
+    # minimal occurrences count to trigger, 3 by default
+    min-occurrences: 3
+  depguard:
+    list-type: blacklist
+    include-go-root: false
+    packages:
+      - github.com/sirupsen/logrus
+      - github.com/golang/protobuf/jsonpb
+      - google.golang.org/protobuf/encoding/protojson
+      - github.com/golang/protobuf/proto
+      - google.golang.org/protobuf/proto
+    packages-with-error-messages:
+      # specify an error message to output when a blacklisted package is used
+      github.com/sirupsen/logrus: "logging is allowed only by logutils.Log"
+      github.com/golang/protobuf/jsonpb: "replace with github.com/gogo/protobuf/jsonpb"
+      google.golang.org/protobuf/encoding/protojson: "replace with github.com/gogo/protobuf/jsonpb"
+      github.com/golang/protobuf/proto: "replace with github.com/gogo/protobuf/proto"
+      google.golang.org/protobuf/proto: "replace with github.com/gogo/protobuf/proto"
+  misspell:
+    # Correct spellings using locale preferences for US or UK.
+    # Default is to use a neutral variety of English.
+    # Setting locale to US will correct the British spelling of 'colour' to 'color'.
+    locale: US
+    ignore-words:
+      - someword
+  lll:
+    # max line length, lines longer will be reported. Default is 120.
+    # '\t' is counted as 1 character by default, and can be changed with the tab-width option
+    line-length: 120
+    # tab width in spaces. Default to 1.
+    tab-width: 1
+  unused:
+    # treat code as a program (not a library) and report unused exported identifiers; default is false.
+    # XXX: if you enable this setting, unused will report a lot of false-positives in text editors:
+    # if it's called for subdir of a project it can't find funcs usages. All text editor integrations
+    # with golangci-lint call it on a directory with the changed file.
+    check-exported: false
+  unparam:
+    # Inspect exported functions, default is false. Set to true if no external program/library imports your code.
+    # XXX: if you enable this setting, unparam will report a lot of false-positives in text editors:
+    # if it's called for subdir of a project it can't find external interfaces. All text editor integrations
+    # with golangci-lint call it on a directory with the changed file.
+    check-exported: false
+  nakedret:
+    # make an issue if func has more lines of code than this setting and it has naked returns; default is 30
+    max-func-lines: 30
```

Comment:
nit; worth making this 1? I think we generally want to discourage naked returns considering we don't really use them anywhere and they're a lot harder to reason about

Reply:
Reducing nesting, typically via early returns, is outlined in Uber's [Go Style Guide](https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting):

> Code should reduce nesting where possible by handling error cases/special conditions first and returning early or continuing the loop. Reduce the amount of code that is nested multiple levels.

Otherwise, we end up with a greater amount of indentation in aggregate, which isn't ideal.